### ISSUE TITLE
On IE11, newlines in code block are not preserved

### ISSUE THREAD

**Issue by kevinzang** [ProseMirror/prosemirror#806]
### Issue details

On IE11, newlines in code block are not preserved.

### Steps to reproduce

Go to prosemirror.net.
In the editor, create a code block containing a newline, e.g.

```
asdf
asdf
```

Exit the code block.
Outside the code block, type some text. The text inside the code block gets newlines replaced with spaces:

```
asdf asdf
```

### ProseMirror version

The version used by prosemirror.net.

### Affected platforms

- [ ] Chrome
- [ ] Firefox
- [x] Internet Explorer (I tried Windows 8.1 IE11 and Windows 10 IE11 via browserstack.com)
- [ ] Other

**Comment by marijnh** [comment]
The issue is probably happening somewhere in `addTextNode` in `model/src/from_dom.js`. Are you able to further debug it?

**Comment by kevinzang** [comment]
I will take a look. Thanks for the hint!

**Comment by kevinzang** [comment]
@marijnh I debugged further.

[prosemirror-view parseBetween](https://github.com/ProseMirror/prosemirror-view/blob/master/src/domchange.js#L112) gets called with the range being the entire doc. We parse with option [preserveWhitespace: $from.parent.type.spec.code ? "full" : true](https://github.com/ProseMirror/prosemirror-view/blob/master/src/domchange.js#L130). It evaluates to `preserveWhitespace: true` since `$from.parent` is the entire doc.

Consequently, in [prosemirror-model addTextNode](https://github.com/ProseMirror/prosemirror-model/blob/master/src/from_dom.js#L379), the `top` node is a code block, yet the `top.options` is `1 (OPT_PRESERVE_WS)`, so [we replace newlines with spaces: value = value.replace(/\r?\n|\r/g, " ")](https://github.com/ProseMirror/prosemirror-model/blob/master/src/from_dom.js#L392).

Basically, the bug is happening because the preserveWhitespace option is not `'full'` when it's supposed to be.

I found a fix: update [NodeViewDesc.parseRule](https://github.com/ProseMirror/prosemirror-view/blob/master/src/viewdesc.js#L557) to pass `preserveWhitespace: this.node.type.spec.code ? 'full' : false`. However, I am not sure if it is a complete fix, as there might be other places where we aren't specifying the expected preserveWhitespace option. For example, [here we have: if (block) this.enter(block)](https://github.com/ProseMirror/prosemirror-model/blob/master/src/from_dom.js#L519). Should it look for a ParseRule corresponding to the top node and pass `rule.preserveWhitespace` similar to what [addElementByRule is doing](https://github.com/ProseMirror/prosemirror-model/blob/master/src/from_dom.js#L449)?

Let me know if there's anything else I can do to help. Thanks!

**Comment by kevinzang** [comment]
I realized this bug can also be reproduced on Chrome by adding the following snippet:

```
// Using https://code.jquery.com/jquery-3.3.1.js
$(document).bind('DOMNodeInserted', function(e) {
    console.log(e.target.innerHTML, $(e.target).find('.nonexistent:not([data-existent])').length);
});
```

See https://glitch.com/edit/#!/cloudy-periodical?path=index.html:14:36.

It seems jQuery lookup via attribute selector (`[data-existent]`) triggers DOM mutation handlers in ProseMirror. I unfortunately am using a third party library, for an unrelated feature, which is doing something to the effect of the snippet above.

**Comment by kevinzang** [comment]
@marijnh Here's a gif showing how to reproduce on Chrome. The glitch is from https://glitch.com/edit/#!/cloudy-periodical?path=index.html:14:36. I'm using Chrome 67.0.3396.79 (Official Build) (64-bit) on MacOS Sierra (10.12.6).

![newline](https://user-images.githubusercontent.com/6496295/41466444-21df34fa-7057-11e8-9996-f27391d320b7.gif)

Basically, select code block, and then alternate between typing "asdf" and "Enter".

**Comment by marijnh** [comment]
Ah, I was typing outside of the code block before, as per the original description. Yes, that does allow me to reproduce it in Chrome (which is _so_ much more comfortable than working with IE11), and I've found the problem. See attached patch, released as part of prosemirror-view 1.3.3

**Comment by kevinzang** [comment]
Thank you! I confirm the patch works (for both the original bug on IE11 and the variant bug on Chrome).
